src/compiler/checker.ts

```diff
@@ -23277,7 +23277,18 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
                     // allow assignments of index types of identical (or similar enough) mapped types.
                     // eg, `keyof {[X in keyof A]: Obj[X]}` should be assignable to `keyof {[Y in keyof A]: Tup[Y]}` because both map over the same set of keys (`keyof A`).
                     // Without this source-side breakdown, a `keyof {[X in keyof A]: Obj[X]}` style type won't be assignable to anything except itself, which is much too strict.
-                    const sourceMappedKeys = nameType && isMappedTypeWithKeyofConstraintDeclaration(mappedType) ? getApparentMappedTypeKeys(nameType, mappedType) : (nameType || getConstraintTypeFromMappedType(mappedType));
+                    let sourceMappedKeys: Type;
+                    if (nameType && isMappedTypeWithKeyofConstraintDeclaration(mappedType)) {
+                        sourceMappedKeys = getApparentMappedTypeKeys(nameType, mappedType);
+                        if (sourceMappedKeys.flags & TypeFlags.Never) {

```

Comment:
> keyof unknown is undefined.

If we think about `keyof T` as a set of types that can be safely used to access properties of `T` in `T[keyof T]` then `never` makes sense here. `never` is the only type that can be safely used as `keyof T` after we instantiate `T` as `unknown`. 

> the fix is looking at the result of getApparentMappedTypeKeys. I'd expect it to be inside getApparentMappedTypeKeys. Isn't there a case where getApparentMappedTypeKeys is never and sourceMappedKeys should stay as never?

That's what I started with but mapping `never` to `PropertyKey` would break the other call site where `getApparentMappedTypeKeys` is used. In there, it's used to get the *target* keys. So this would remove the error below:

```ts
type StringKeys2<T> = keyof {
  [P in keyof T as T[P] extends string ? P : never]: any;
};
function h<T>(z: StringKeys2<T>) {
  z = "foo"; // error
}
```

Reply:
This all has me thinking that the original code that uses `getApparentMappedTypeKeys` on the source side is just... wrong?